go-controller/pkg/crd/adminpolicybasedroute/v1/types.go

```diff
@@ -0,0 +1,125 @@
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:resource:path=adminpolicybasedexternalroute,scope=Cluster
+// +kubebuilder:object:root=true
+// +kubebuilder:subresource:status
+
+// AdminPolicyBasedExternalRoute is the Schema for the AdminPolicyBasedExternalRoutes API
+type AdminPolicyBasedExternalRoute struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	Spec              AdminPolicyBasedExternalRouteSpec `json:"spec,omitempty"`
+	Status            AdminPolicyBasedRouteStatus       `json:"status,omitempty"`
+}
+
+// AdminPolicyBasedExternalRouteSpec defines the desired state of AdminPolicyBasedExternalRoute
+type AdminPolicyBasedExternalRouteSpec struct {
+	Policies []*ExternalPolicy `json:"policies"`
+}
+
+type ExternalPolicy struct {
+	From     ExternalNetworkSource `json:"from"`
+	NextHops ExternalNextHops      `json:"nextHops"`
+}
+
+type ExternalNetworkSource struct {
+	NamespaceSelector metav1.LabelSelector `json:"namespaceSelector,omitempty"`
+}
+
+type ExternalNextHops struct {
+	// +kubebuilder:validation:UniqueItems=true
+	StaticHops []*StaticHop `json:"static,omitempty"`
+	// +kubebuilder:validation:UniqueItems=true
+	DynamicHops []*DynamicHop `json:"dynamic,omitempty"`
+}
+
+type StaticHop struct {
+	IP           string `json:"ip,omitempty"`
+	BFDEnabled   bool   `json:"bfdEnabled,omitempty"`
+	SkipHostSNAT bool   `json:"skipHostSNAT,omitempty"`
+}
+
+type DynamicHop struct {
+	PodSelector           metav1.LabelSelector  `json:"podSelector,omitempty"`
+	NamespaceSelector     *metav1.LabelSelector `json:"namespaceSelector,omitempty"`
+	NetworkAttachmentName string                `json:"networkAttachmentName,omitempty"`
+	BFDEnabled            bool                  `json:"bfdEnabled,omitempty"`
+	SkipHostSNAT          bool                  `json:"skipHostSNAT,omitempty"`
+}
+
+// +kubebuilder:object:root=true
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// AdminPolicyBasedExternalRouteList contains a list of AdminPolicyBasedExternalRoutes
+type AdminPolicyBasedExternalRouteList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []AdminPolicyBasedExternalRoute `json:"items"`
+}
+
+// AdminPolicyBasedRouteStatus describes the current status of the AdminPolicyBased route types.
+// +kubebuilder:object:root=true
+type AdminPolicyBasedRouteStatus struct {
+	// An array of Human-readable messages indicating details about the status of the object.
```

Comment:
IMHO bad API design, controller can put whatever it wants? that too an array of them ? how will people correlate these to each of the values in the object? eg:
staticHop[0] -> goodCondition
staticHop[1] -> badCondition
dynamicHop[0] -> not configured yet, still retrying

and so on.. How can you reflect that using the status? there is nothing tying the orders here right?

Reply:
It doesn't work like that: the status is a summary of the changes applied as a whole to the CR, it does not reflect individual hops. The `Status` field within reflects the latest condition for the CR, in this case I capture 2 types: `Success` or `Failed`. The former states that the CR was applied without errors, while the later informs there were errors that can be found in the latest entry in the `messages` field.